packages/hdwallet-native/src/crypto/EncryptedWallet.ts

```diff
@@ -0,0 +1,134 @@
+import { validateMnemonic } from "bip39";
+import { CryptoEngine } from "./engines";
+import CryptoHelper from "./CryptoHelper";
+import { CipherString, SymmetricCryptoKey } from "./classes";
+import * as utils from "./utils";
+
+export class EncryptedWallet {
+  readonly #engine: CryptoEngine;
+  readonly #helper: CryptoHelper;
+  #deviceId: string;
+  #email: string;
+  #encryptedWallet: string;
+  #key: SymmetricCryptoKey;
+  #password: string;
+
+  constructor(engine: CryptoEngine) {
+    if (!engine) {
+      throw new Error("Missing cryptography engine");
+    }
+    this.#engine = engine;
+    this.#helper = new CryptoHelper(this.#engine);
+  }
+
+  /**
+   * Whether or not the wallet has been initialized with an email and password
+   */
+  get isInitialized() {
+    return Boolean(this.#email && this.#password && this.#key);
+  }
+
+  /**
+   * The email provided to the `init` function
+   */
+  get email() {
+    return this.#email;
+  }
+
+  /**
+   * A hash derived from the email and password to be used for authentication
+   */
+  get passwordHash() {
+    return this.#key?.hashKeyB64;
+  }
+
+  /**
+   * A string representation of the encrypted seed phrase
+   */
+  get encryptedWallet() {
+    return this.#encryptedWallet;
+  }
+
+  /**
+   * Get an ID based on the mnemonic
+   * Calling `decrypt` will set this value after decryption is successful
+   */
+  get deviceId() {
+    return this.#deviceId;
+  }
+
+  /**
+   * Set the encrypted wallet by providing a string representation
+   * @throws {Error} throws if `wallet` is not a valid encrypted wallet string
+   */
+  set encryptedWallet(wallet: string) {
+    this.#encryptedWallet = new CipherString(wallet).encryptedString;
+  }
+
+  /**
+   * Initialize the wallet with and email and password
+   *
+   * This cannot be done in the constructor because it performs async operations
+   */
+  async init(email: string, password: string, encryptedWallet?: string): Promise<EncryptedWallet> {
+    if (!(email && typeof email === "string" && email.length > 0)) {
+      throw new Error("Invalid email address");
+    }
+    if (!(password && typeof password === "string" && password.length > 0)) {
+      throw new Error("Invalid password");
+    }
+
+    this.#email = email.normalize("NFKC").trim().toLowerCase();
+    this.#password = password.normalize("NFKC");
+    this.#key = await this.#helper.makeKey(this.#password, this.#email);
+
+    if (encryptedWallet) {
+      this.encryptedWallet = encryptedWallet;
```

Comment:
This assignment could be done in the constructor right?

Reply:
just the encrypted wallet part? Right now the constructor only takes in the engine then requires an init call. `encryptedWallet` by itself isn't useful because you need the email/password to do anything with it anyway